zrenderer/geometry/zrbvh/rbvhtree.cpp

```diff
@@ -0,0 +1,586 @@
+        float n = transformed.min().z(), f = transformed.max().z();
+
+        Matrix4f projMatrix = Utils::orthoProjection( l, r, t, b, -n, -f );
+        Matrix4f m = projMatrix * viewMatrix;
+        _transform = m.block( 0, 0, 3, 4 );
+    }
+
+    void rasterize( oglplus::Uniform<oglplus::Mat4x3f>& proj )
+    {
+        oglplus::Renderbuffer heightmap;
+        oglplus::Renderbuffer depth;
+        oglplus::Framebuffer fbo;
+        oglplus::Context gl;
+
+        uint32_t sz = (uint32_t)_R;
+
+        gl.Bound( oglplus::Renderbuffer::Target::Renderbuffer, heightmap )
+            .Storage( oglplus::InternalFormat::R32F, sz, sz );
+
+
+        gl.Bound( oglplus::Renderbuffer::Target::Renderbuffer, depth )
+            .Storage( oglplus::InternalFormat::DepthComponent32, sz, sz );
+
+        gl.Bound( oglplus::Framebuffer::Target::Draw, fbo )
+            .AttachRenderbuffer( oglplus::FramebufferAttachment::Color, heightmap )
+            .AttachRenderbuffer( oglplus::FramebufferAttachment::Depth, depth )
+            .Complete();
+
+        proj.Set( oglplus::Mat4x3f( _transform.data(), 12 ) );
+
+        gl.Viewport( sz, sz );
+        gl.ClearColor( 0.5, 0.5f, 0.5f, 0.5f );
+        gl.Clear().ColorBuffer().DepthBuffer();
+        gl.DrawElements(
+            oglplus::PrimitiveType::Triangles,
+            _count * 3,
+            (uint32_t*)nullptr + _startIdx*3);
+
+        gl.Bound( oglplus::Framebuffer::Target::Read, fbo );
+        gl.ReadBuffer( oglplus::FramebufferColorAttachment::_0 );
+
+        _heightmap.resize( sz*sz );
+        gl.ReadPixels( 0, 0, sz, sz, oglplus::PixelDataFormat::Red,
+            oglplus::PixelDataType::Float, _heightmap.data() );
+        gl;
+    }
+
+    RBVHTrianglesPtr _triangles;
+    RBVHTriangleIndicesPtr _triangleIndices;
+    float _R, _rate;
+    Vector3f _sumAreaWeightedNormal;
+};
+
+struct RBVHNodeData
+{
+    uint32_t _data;            // 4 bytes: 3 bytes Idx, 6 bits triangle_count, 2 bits type
+    AlignedBox3h _boundingBox; // 12 bytes
+};
+
+struct RBVHTree::Impl
+{
+    Impl( ConstMeshPtr mesh, float rho, float alpha, float upperLimit )
+    {
+        MeshPtr trimesh;
+        RBVHNode nodes( mesh, trimesh, rho, alpha, upperLimit );
+
+        nodes.traverse([this]( RBVHNode* node )
+            {
+                _nodes.push_back( getNodeData( node ) );
+                if( node->_type == NT_HEIGHTMAP )
+                {
+                    _transforms.push_back( node->_transform );
+                    _heightmaps.push_back( node->_heightmap );
+                }
+            } );
+
+        _vertices = std::move( trimesh->getVertices() );
+        _indices = std::move( trimesh->getFaces() );
+
+    }
+
+    RBVHNodeData getNodeData( RBVHNode* node )
+    {
+        uint32_t data = node->_type;
+
+        switch( node->_type )
+        {
+        case NT_TRIANGLES:
+            data |= ((node->_startIdx << 8) & 0xFFFFFF00);
```

Comment:
bitfield lar için https://github.com/BlueBrain/Livre/blob/master/livre/core/data/NodeId.h